File: problems/simon-ids-continuity/compute/q1/verify_frontier.py
#!/usr/bin/env python3
"""Verify the exact Bourgain--Klein frontier and free-direction powers."""

from __future__ import annotations

import csv
import sys
from fractions import Fraction
from pathlib import Path


def read_rows(path: Path) -> list[dict[str, str]]:
    with path.open(newline="", encoding="utf-8") as handle:
        return list(csv.DictReader(handle))


def shown(value: Fraction) -> str:
    return str(value.numerator) if value.denominator == 1 else str(value)


def verify_frontier(root: Path) -> list[str]:
    rows = read_rows(root / "frontier_certificate.csv")
    dimensions = [int(row["dimension"]) for row in rows]
    if dimensions != list(range(2, 9)):
        raise ValueError(f"unexpected dimension range: {dimensions}")

    report = ["Bourgain-Klein frontier", "d,gap,kappa,works"]
    for row in rows:
        dimension = int(row["dimension"])
        expected_gap = Fraction(4 - dimension, 3 * (dimension - 1))
        expected_kappa = Fraction(4 - dimension, 8)
        expected_works = dimension < 4

        stored_gap = Fraction(int(row["gap_num"]), int(row["gap_den"]))
        stored_kappa = Fraction(
            int(row["kappa_num"]), int(row["kappa_den"])
        )
        stored_works = row["works"] == "true"
        if (stored_gap, stored_kappa, stored_works) != (
            expected_gap,
            expected_kappa,
            expected_works,
        ):
            raise ValueError(f"frontier mismatch at d={dimension}")

        report.append(
            f"{dimension},{shown(expected_gap)},{shown(expected_kappa)},"
            f"{str(expected_works).lower()}"
        )
    return report


def verify_free_directions(root: Path) -> list[str]:
    rows = read_rows(root / "free_direction_certificate.csv")
    dimensions = [int(row["free_dimensions"]) for row in rows]
    if dimensions != list(range(1, 9)):
        raise ValueError(f"unexpected free-dimension range: {dimensions}")

    report = ["free-direction modulus", "m,alpha,theta"]
    for row in rows:
        dimension = int(row["free_dimensions"])
        expected_alpha = Fraction(dimension, 2)
        expected_modulus = min(Fraction(1), expected_alpha)
        stored_alpha = Fraction(int(row["alpha_num"]), int(row["alpha_den"]))
        stored_modulus = Fraction(
            int(row["modulus_num"]), int(row["modulus_den"])
        )
        if (stored_alpha, stored_modulus) != (
            expected_alpha,
            expected_modulus,
        ):
            raise ValueError(f"free-direction mismatch at m={dimension}")

        report.append(
            f"{dimension},{shown(expected_alpha)},{shown(expected_modulus)}"
        )
    return report


def main() -> None:
    root = Path(sys.argv[1]) if len(sys.argv) == 2 else Path(__file__).parent
    report = verify_frontier(root) + verify_free_directions(root)
    report.append("certificate verified")
    print("\n".join(report))


if __name__ == "__main__":
    main()

File: problems/simon-ids-continuity/compute/q1/verify_frontier.rs
use std::env;
use std::fmt;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
struct Rational {
    numerator: i64,
    denominator: i64,
}

fn gcd(mut left: i64, mut right: i64) -> i64 {
    left = left.abs();
    right = right.abs();
    while right != 0 {
        let remainder = left % right;
        left = right;
        right = remainder;
    }
    if left == 0 { 1 } else { left }
}

impl Rational {
    fn new(mut numerator: i64, mut denominator: i64) -> Self {
        assert!(denominator != 0, "zero denominator");
        if denominator < 0 {
            numerator = -numerator;
            denominator = -denominator;
        }
        let divisor = gcd(numerator, denominator);
        Self {
            numerator: numerator / divisor,
            denominator: denominator / divisor,
        }
    }
}

impl fmt::Display for Rational {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.denominator == 1 {
            write!(formatter, "{}", self.numerator)
        } else {
            write!(formatter, "{}/{}", self.numerator, self.denominator)
        }
    }
}

fn csv_rows(path: &Path) -> Vec<Vec<String>> {
    let contents = fs::read_to_string(path)
        .unwrap_or_else(|error| panic!("cannot read {}: {error}", path.display()));
    contents
        .lines()
        .skip(1)
        .filter(|line| !line.trim().is_empty())
        .map(|line| line.split(',').map(str::to_owned).collect())
        .collect()
}

fn integer(field: &str) -> i64 {
    field
        .parse::<i64>()
        .unwrap_or_else(|error| panic!("bad integer {field}: {error}"))
}

fn verify_frontier(root: &Path, report: &mut Vec<String>) {
    let rows = csv_rows(&root.join("frontier_certificate.csv"));
    assert_eq!(rows.len(), 7, "frontier row count");
    report.push("Bourgain-Klein frontier".to_owned());
    report.push("d,gap,kappa,works".to_owned());

    for (index, row) in rows.iter().enumerate() {
        assert_eq!(row.len(), 6, "frontier column count");
        let dimension = integer(&row[0]);
        assert_eq!(dimension, index as i64 + 2, "dimension sequence");

        let stored_gap = Rational::new(integer(&row[1]), integer(&row[2]));
        let stored_kappa = Rational::new(integer(&row[3]), integer(&row[4]));
        let stored_works = match row[5].as_str() {
            "true" => true,
            "false" => false,
            other => panic!("bad Boolean {other}"),
        };

        let gap = Rational::new(4 - dimension, 3 * (dimension - 1));
        let kappa = Rational::new(4 - dimension, 8);
        let works = dimension < 4;
        assert_eq!(stored_gap, gap, "gap at d={dimension}");
        assert_eq!(stored_kappa, kappa, "kappa at d={dimension}");
        assert_eq!(stored_works, works, "outcome at d={dimension}");
        report.push(format!("{dimension},{gap},{kappa},{works}"));
    }
}

fn verify_free_directions(root: &Path, report: &mut Vec<String>) {
    let rows = csv_rows(&root.join("free_direction_certificate.csv"));
    assert_eq!(rows.len(), 8, "free-direction row count");
    report.push("free-direction modulus".to_owned());
    report.push("m,alpha,theta".to_owned());

    for (index, row) in rows.iter().enumerate() {
        assert_eq!(row.len(), 5, "free-direction column count");
        let dimension = integer(&row[0]);
        assert_eq!(dimension, index as i64 + 1, "free-dimension sequence");

        let stored_alpha = Rational::new(integer(&row[1]), integer(&row[2]));
        let stored_modulus = Rational::new(integer(&row[3]), integer(&row[4]));
        let alpha = Rational::new(dimension, 2);
        let modulus = if dimension < 2 {
            alpha
        } else {
            Rational::new(1, 1)
        };
        assert_eq!(stored_alpha, alpha, "alpha at m={dimension}");
        assert_eq!(stored_modulus, modulus, "modulus at m={dimension}");
        report.push(format!("{dimension},{alpha},{modulus}"));
    }
}

fn main() {
    let arguments: Vec<String> = env::args().collect();
    let root = if arguments.len() == 2 {
        PathBuf::from(&arguments[1])
    } else if arguments.len() == 1 {
        PathBuf::from(".")
    } else {
        panic!("usage: verify_frontier [certificate-directory]");
    };

    let mut report = Vec::new();
    verify_frontier(&root, &mut report);
    verify_free_directions(&root, &mut report);
    report.push("certificate verified".to_owned());
    println!("{}", report.join("\n"));
}
